crates/gpui/src/platform/windows/platform.rs

```diff
@@ -0,0 +1,304 @@
+    time::Duration,
+};
+
+use anyhow::{anyhow, Result};
+use async_task::Runnable;
+use futures::channel::oneshot::Receiver;
+use parking_lot::Mutex;
+use time::UtcOffset;
+use util::SemanticVersion;
+use windows::Win32::{
+    Foundation::HWND,
+    UI::WindowsAndMessaging::{
+        DispatchMessageW, PeekMessageW, PostQuitMessage, TranslateMessage, MSG, PM_REMOVE, WM_QUIT,
+    },
+};
+
+use crate::{
+    Action, AnyWindowHandle, BackgroundExecutor, ClipboardItem, CursorStyle, ForegroundExecutor,
+    Keymap, Menu, PathPromptOptions, Platform, PlatformDisplay, PlatformInput, PlatformTextSystem,
+    PlatformWindow, Task, WindowAppearance, WindowOptions, WindowsDispatcher, WindowsDisplay,
+    WindowsTextSystem, WindowsWindow,
+};
+
+pub(crate) struct WindowsPlatform {
+    inner: Rc<WindowsPlatformInner>,
+}
+
+pub(crate) struct WindowsPlatformInner {
+    background_executor: BackgroundExecutor,
+    pub(crate) foreground_executor: ForegroundExecutor,
+    main_receiver: flume::Receiver<Runnable>,
+    text_system: Arc<WindowsTextSystem>,
+    callbacks: Mutex<Callbacks>,
+    pub(crate) window_handles: RefCell<HashSet<AnyWindowHandle>>,
+}
+
+#[derive(Default)]
+struct Callbacks {
+    open_urls: Option<Box<dyn FnMut(Vec<String>)>>,
+    become_active: Option<Box<dyn FnMut()>>,
+    resign_active: Option<Box<dyn FnMut()>>,
+    quit: Option<Box<dyn FnMut()>>,
+    reopen: Option<Box<dyn FnMut()>>,
+    event: Option<Box<dyn FnMut(PlatformInput) -> bool>>,
+    app_menu_action: Option<Box<dyn FnMut(&dyn Action)>>,
+    will_open_app_menu: Option<Box<dyn FnMut()>>,
+    validate_app_menu_command: Option<Box<dyn FnMut(&dyn Action) -> bool>>,
+}
+
+impl WindowsPlatform {
+    pub(crate) fn new() -> Self {
+        let (main_sender, main_receiver) = flume::unbounded::<Runnable>();
+        let dispatcher = Arc::new(WindowsDispatcher::new(main_sender));
+        let background_executor = BackgroundExecutor::new(dispatcher.clone());
+        let foreground_executor = ForegroundExecutor::new(dispatcher);
+        let text_system = Arc::new(WindowsTextSystem::new());
+        let callbacks = Mutex::new(Callbacks::default());
+        let window_handles = RefCell::new(HashSet::new());
+        let inner = Rc::new(WindowsPlatformInner {
+            background_executor,
+            foreground_executor,
+            main_receiver,
+            text_system,
+            callbacks,
+            window_handles,
+        });
+        Self { inner }
+    }
+}
+
+impl Platform for WindowsPlatform {
+    fn background_executor(&self) -> BackgroundExecutor {
+        self.inner.background_executor.clone()
+    }
+
+    fn foreground_executor(&self) -> ForegroundExecutor {
+        self.inner.foreground_executor.clone()
+    }
+
+    fn text_system(&self) -> Arc<dyn PlatformTextSystem> {
+        self.inner.text_system.clone()
+    }
+
+    fn run(&self, on_finish_launching: Box<dyn 'static + FnOnce()>) {
+        on_finish_launching();
+        loop {
```

Comment:
currently the event loop is just spinning without any blocking, we might wanna change it to block on events (using `WaitMessage` or `GetMessageW`).

But if we do block then foreground executor will have to change, it will need to create an invisible window and schedule events to that window to wake up the event loop.

Reply:
Though IMO that can be done in later PRs.

Reply:
Can something like `WM_ENTERIDLE` be used for processing the foreground tasks?
Having a busy thread without sleeping would be unfortunate.

Reply:
I think that we could use `MsgWaitForMultipleObject` to wait for window messages and signals from background threads at the same time, or use `PostThreadMessage` and raw pointers instead of sender and receiver, so that we can wait on `GetMessage`.

Reply:
This is now changed to use `MsgWaitForMultipleObjects`